chore(journal): split journal record into metadata and indexed record

deepthidevaki opened the pull request:
## Description

This is a PR in preparation to #6423 . In this PR 

* Splits journal record into a JournalRecordMetadata and JournalIndexedRecord. The indexed record consists of index, asqn and the data of the record. The metadata consists of the checksum and the length of the serialized indexed record. They are serialized and written separately. 
* Extracts the serialization logic. A KryoSerializer implements the serialization logic. When SBE is introduced, it is only required to replace the KryoSerializer with an SBE serializer. 

## Related issue

Related to #6423 

## Definition of Done

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [ ] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/zeebe-io/zeebe/compare/stable/0.24...develop?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/0.25`) to the PR, in case that fails you need to create backports manually.

Testing:
* [x] There are unit/integration tests that verify all acceptance criterias of the issue
* [ ] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark 

Documentation: 
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] New content is added to the [release announcement](https://drive.google.com/drive/u/0/folders/1DTIeswnEEq-NggJ25rm2BsDjcCQpDape)

npepinpe commented on journal/pom.xml:
```diff
@@ -33,11 +33,6 @@
       <artifactId>simpleclient</artifactId>
     </dependency>
 
-    <dependency>
```
I'm confused how this works - we remove the dependency but we implemented `KryoSerializer`? How is this not causing a build problem? :thinking:

deepthidevaki replied:
KryoSerializer does not neeed Kryo, but it serializes via Namespaces from atomix-utils. I don't know why we needed this dependency before.

deepthidevaki replied:
>     1. I would rename `JournalIndexedRecord` to `RecordData`, such that a `JournalRecord` is the combination of the metadata and the data? It seems more natural to me.

:+1: 
>     2. In the `MappedJournalSegmentWriter`, we often pass the `buffer` field around as a method argument, but I'm not sure why - do we sometimes expect a different one? Not that I could find, but maybe I missed something.

:+1:

>     3. If it's possible, I'd let the serializers decide how to serialize the data for each message type directly. By this I mean, instead of writing `JournalRecordMetadataImpl`, I would let it write the metadata properties, e.g. `serializer.write(writeBuffer, checksum)`, `serializer.write(writeBuffer, index, asqn, data)`, etc. Then we only need concrete classes for reading back, not for writing. I get that it doesn't scale if we have lots of params, but it's probably fine for now? wdyt? Are these classes currently needed because we use Kryo? Even then I would consider these as implementation details of the serializer, no?

We can remove the interface and only keep the implementation classes. We are using the same implementation class for both reading and writing so there is actually no need to have an interface for it.  If we have only the implementation classes, would you then like to change the serializer api? What is the advantage of it?

npepinpe replied:
Ah! Makes sense :smile:

npepinpe commented on journal/src/main/java/io/zeebe/journal/file/MappedJournalSegmentWriter.java:
```diff
@@ -88,9 +94,31 @@ public JournalRecord append(final long asqn, final DirectBuffer data) {
-    final int recordStartPosition = buffer.position();
-    lastEntry = write(buffer, recordIndex, asqn, data);
-    index.index(lastEntry, recordStartPosition);
+    final MutableDirectBuffer writeBuffer = new UnsafeBuffer();
```
We will need a follow up issue to clean this method up - there's several improvements we could do (mostly always using a `MutableDirectBuffer`, safely managing the position/limit, etc.) but I guess it's out of scope.

npepinpe commented on journal/src/main/java/io/zeebe/journal/file/MappedJournalSegmentWriter.java:
```diff
@@ -105,72 +133,97 @@ public void append(final JournalRecord record) {
+  private void checkCanWrite(final ByteBuffer buffer, final JournalIndexedRecord indexedRecord) {
+    final var recordOffset = serializer.getMetadataLength();
+
+    final int estimatedRecordLength =
```
If we're estimating, then we probably want to estimate greater than or equal to the real size right? Is it what we're doing now? If we don't, then we could estimate something smaller than the length and fail to write it right? But I guess at worse we just would write in the next segment, so...

deepthidevaki replied:
KryoSerializer throws BufferOverflowException if there is not enough space to write in this segment. This estimated length is mostly only used to check with the maxEntrySize which I guess we want to remove it. So this is only a temporary state.

npepinpe commented on journal/src/main/java/io/zeebe/journal/file/MappedJournalSegmentWriter.java:
```diff
@@ -105,72 +133,97 @@ public void append(final JournalRecord record) {
+      final int offset,
+      final JournalIndexedRecord indexedRecord) {
+    writeBuffer.wrap(buffer, offset, buffer.limit() - offset);
+    final var recordLength = serializer.write(indexedRecord, writeBuffer);
+    return recordLength;
```
Can just return immediately here

```suggestion
    return serializer.write(indexedRecord, writeBuffer);
```

npepinpe commented on journal/src/main/java/io/zeebe/journal/file/MappedJournalSegmentWriter.java:
```diff
@@ -105,72 +133,97 @@ public void append(final JournalRecord record) {
-      final ByteBuffer buffer, final long index, final long asqn, final DirectBuffer data) {
+    final JournalIndexedRecord indexedRecord =
+        new JournalIndexedRecordImpl(record.index(), record.asqn(), record.data());
+    checkCanWrite(buffer, indexedRecord);
```
It would be nice in the future if for expected exceptions we don't throw but rather handle these error cases properly. I get that's out of scope, but something we should take care of in the future.

npepinpe commented on journal/src/main/java/io/zeebe/journal/JournalRecord.java:
```diff
@@ -38,7 +38,7 @@
    * @return checksum
    */
-  int checksum();
+  long checksum();
```
:+1:

npepinpe commented on journal/src/main/java/io/zeebe/journal/file/record/JournalRecordSerializer.java:
```diff
@@ -0,0 +1,84 @@
+import org.agrona.DirectBuffer;
+import org.agrona.MutableDirectBuffer;
+
+public interface JournalRecordSerializer {
```
I'm not sure about this interface:

1. Like I mentioned, I don't think having interfaces/implementation classes for the data to write is a good idea. It simplifies the method signature I guess, but then implementation also needs the impl classes? Why not have only impl classes? If it's only to simplify the signature, then for now I'd start by passing the fields directly in the various write methods.
2. We might want to consider later also passing the write buffer/read buffer + the offset/length to avoid allocating buffer views all the time. It's also a common pattern we use elsewhere and which the Agrona buffers support nicely.
3. Why not simply have `writeMetadata` and `writeData` as different methods? I'm not sure if the overload is useful here.
4. `getMetadataLength()` makes me realize we should think about what we want - is how the metadata/data laid out implementation details of the serializer, or is it the writer, and the serializer just serializes the messages to the given buffer? If it's implementation detail, then you just say `int length = serializer.writeData(buffer, index, asqn, data)` and it will already leave space for the metadata, and then `serializer.writeMetadata(buffer, checksum, length)` will write at the proper offset. Then the writer doesn't know if metadata is first, or after, etc, and we're coupling both more tightly I guess...idk, wdyt? But then you don't need `getMetadataLength()` for example. That also means with Kryo you wouldn't need to estimate size, I think?
5. I'm not sure if we need `hasMetadata` - maybe now, but later when we have the magic header, I guess we won't right?

deepthidevaki replied:
1. You are right. We don't need an interface and implementation classes for the data. Initially, we had different implementation for reading and writing. But that is not the case anymore. So it would make sense to only have the implementation classes.
2. Make sense. Do you want to address it in this PR?
3. Agree.
4. The writer and reader needs to know where the data is written so that it can calculate the checksum. If it is hidden in the serializer, we are back to the problem we discussed in the draft PR. 
5. When we have magic header we don't need `hasMetadata`.

npepinpe replied:
Re 4, you're right, I forgot - the reader needs to know where the offset of the data starts. I guess we need something like we have for the `LoggedEvent`, where we have `LoggedEvent#getValueBuffer()` and `LoggedEvent#getValueOffset()` (and same for the metadata). I guess this is fine for now, it still seems strange to me but maybe it's clearer what's the right API when we're mostly done.

npepinpe reviewed (changes requested):
:+1: 

Some notes:

1. I would rename `JournalIndexedRecord` to `RecordData`, such that a `JournalRecord` is the combination of the metadata and the data? It seems more natural to me.
2. In the `MappedJournalSegmentWriter`, we often pass the `buffer` field around as a method argument, but I'm not sure why - do we sometimes expect a different one? Not that I could find, but maybe I missed something.
3. If it's possible, I'd let the serializers decide how to serialize the data for each message type directly. By this I mean, instead of writing `JournalRecordMetadataImpl`, I would let it write the metadata properties, e.g. `serializer.write(writeBuffer, checksum)`, `serializer.write(writeBuffer, index, asqn, data)`, etc. Then we only need concrete classes for reading back, not for writing. I get that it doesn't scale if we have lots of params, but it's probably fine for now? wdyt? Are these classes currently needed because we use Kryo? Even then I would consider these as implementation details of the serializer, no?


So that was more comments than I thought :sweat_smile: Happy to go through it f2f

deepthidevaki commented on journal/src/main/java/io/zeebe/journal/file/ChecksumGenerator.java:
```diff
@@ -15,19 +15,17 @@
 import java.util.zip.CRC32;
-import org.agrona.DirectBuffer;
 
 public final class ChecksumGenerator {
```
Actually the goal was to extract the common logic from the reader and writer to this class - i.e slicing the buffer or calculate the checksum from the DirectBuffer etc. The current implementation doesn't do it. So we can either remove this class or extract the common parts from the reader and writer to this class.

npepinpe replied:
Let's create a follow up issue to do this, I think it's fine that it takes in a `ByteBuffer`, but we might want to change the signature to something like:

```java
private final Checksum checksum;

public long compute(final ByteBuffer, final int offset, final int length) {
  buffer.mark();

  try {
    checksum.update(buffer.position(offset).limit(offset + length);
    return checksum.getValue();
  } finally {
    buffer.reset();
    checksum.reset();
  }
}
```

We can do that at the end though once things are a little clearer.

deepthidevaki commented:
>    I would rename `JournalIndexedRecord` to `RecordData`, such that a `JournalRecord` is the combination of the metadata and the data? It seems more natural to me.

:+1: 
>   In the `MappedJournalSegmentWriter`, we often pass the `buffer` field around as a method argument, but I'm not sure why - do we sometimes expect a different one? Not that I could find, but maybe I missed something.

:+1:

>    If it's possible, I'd let the serializers decide how to serialize the data for each message type directly. By this I mean, instead of writing `JournalRecordMetadataImpl`, I would let it write the metadata properties, e.g. `serializer.write(writeBuffer, checksum)`, `serializer.write(writeBuffer, index, asqn, data)`, etc. Then we only need concrete classes for reading back, not for writing. I get that it doesn't scale if we have lots of params, but it's probably fine for now? wdyt? Are these classes currently needed because we use Kryo? Even then I would consider these as implementation details of the serializer, no?

We can remove the interface and only keep the implementation classes. We are using the same implementation class for both reading and writing so there is actually no need to have an interface for it.  If we have only the implementation classes, would you then like to change the serializer api? What is the advantage of it?

npepinpe commented:
I don't actually expect we will even need the implementation classes for the SBE one right? It seems to me that these implementation classes are only used to temporarily group the required arguments together and pass them on to the serializer no? Do we use them for anything else?

deepthidevaki commented:
> I don't actually expect we will even need the implementation classes for the SBE one right? It seems to me that these implementation classes are only used to temporarily group the required arguments together and pass them on to the serializer no? Do we use them for anything else?

We need for it reading. We read the metadata and the data separately.

deepthidevaki commented:
@npepinpe  Updated the PR:
1. Renamed JournalIndexedRecord and JournalRecordMetadata to `RecordData` and `RecordMetadata`
2. Removed the Interfaces for `RecordData` and `RecordMetadata` and only keep the concrete classes. The same class is used for both reading and writing.
3. Updated ChecksumGenerator interface.
4. Remove creating DirectBuffer object on each append. Instead the directbuffer wraps the entire buffer. Serializer accepts the offset for writing and reading.

npepinpe commented on journal/src/main/java/io/zeebe/journal/file/ChecksumGenerator.java:
```diff
@@ -23,9 +23,10 @@
+    buffer.position(offset);
     crc32.reset();
-    crc32.update(buffer);
+    crc32.update(buffer.slice().limit(length));
```
As this is shared by both reader/writer, and we don't want to mutate, I would propose using `asReadOnlyBuffer` instead of slice - it means you need to set the position after, e.g. `buffer.asReadOnlyBuffer().position(offset).limit(offset + limit)`

IIRC limit is the max offset, not the length :sweat_smile:

npepinpe commented on journal/src/main/java/io/zeebe/journal/file/record/RecordData.java:
```diff
@@ -17,35 +17,32 @@
   @Override
   public String toString() {
-    return "JournalIndexedRecordImpl{" + "index=" + index + ", asqn=" + asqn + '}';
+    return "RecordData{" + "index=" + index + ", asqn=" + asqn + ", data=" + data + '}';
```
Lately I've been using the `PrintBufferUtil` from Agrona here to get the hexDump with a max length - I still don't know if that's the best thing in general (maybe security concerns? though outputting the raw bytes is the same I guess), but it was very useful when debugging.

deepthidevaki replied:
Should we output data? I would say `data` can be removed from `toString`.

npepinpe replied:
Probably makes sense to remove it for now.

npepinpe reviewed (approved):
Looks good!

Some notes, mostly thoughts but they could all be addressed as follow up issues:

- The more I think about it, the more I think we should just have type serializers instead of a general serializer...anyway I'm probably overthinking it.
- There's some other things but I think it makes sense to review this once SBE is in since that will likely also give us a better idea of how we will actually use everything. It would also open up some new things, like maybe adopting the same streaming approach to serialization as SBE recommends for deserialization.

deepthidevaki commented:
bors r+
